### ISSUE TITLE
[feature request] Bit packed tensors

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#32867]
A usecase: storing a full backtracking pointer matrix can be okay for needleman/ctc alignment (4x memory saving compared to uint8 representation), if 2bit data type is used. Currently it's possible to do this with bit manipulation magic, but probably not very efficient (store and load will require masking and shifting, not fused)

Another usecase: compressed BoolTensor for boolean networks

Another usecase: low-bit quantized representations.

Is something like this already implemented for quantization? Probably a simple version of this feature could be providing some explicitly utility functions like calculating size of the holder `uint8` tensor, fused store and load functions (potentially explicitly batched, e.g. actual store is delayed until some aligned number of memory lines has arrived)

**Comment by ezyang** [comment]
One difficulty is that in many points of our code we assume all tensor elements are addressable (for example, for views), and this would not be the case with bit-packed tensors.

**Comment by vadimkantorov** [comment]
I wonder if we could design some explicit pack/unpack/load/store/index util methods that would be enough for basic usage (like numpy does with packbits/unpackbits)

**Comment by vadimkantorov** [comment]
A simple interface could be packbits / unpackbits like in NumPy with additional bitness argument (to support 1-bit, 2-bit and 4-bit) and dim argument. It should maybe support out argument for unpacked uint8 tensor. Unpacked dimension could always be a new zeroth dimension.

**Comment by vadimkantorov** [comment]
https://www.microsoft.com/en-us/research/uploads/prod/2018/02/KoPhiliposeTashevZarar_ICASSP_2018.pdf suggests that XNOR and POPCNT functionality is useful for 1-bit networks

**Comment by vadimkantorov** [comment]
Arguments that can be helpful for packbits/unpackbits:
1) `mask` - a bit mask integer, specifying pack/unpack compress mask like in [compress](https://www.officedaytime.com/simd512e/simdimg/compress.php?f=vpcompressq) [expand](https://www.officedaytime.com/simd512e/simdimg/expand.php?f=vexpandq) instructions -> this is slightly more flexible than a single `nbits=1|2|4` arg)

2) dim -> packing / unpacking along a given dim (during unpacking it can then only be done across an already existing dim, maybe that's fine for dense dimensions)

3) target dim size may be needed for unpacking to undo the padding

4) out argument

I guess on packbits/unpackbits can be implemented with those compress/expand SIMD instructions.

**Comment by vadimkantorov** [comment]
in my code I'd do sth like: `torch.packbits(something.argmax(dim = -1), mask = 0b11, dim  =-1, out = my_uint8_array[k])`

**Comment by vadimkantorov** [comment]
I think one can think about them as surfacing compress/expand SIMD functionality to user land and reimplementing it on GPU

**Comment by gchanan** [comment]
pack and unpack seem worth doing.  The other parts (i.e. compress/expand) could be useful, but I'm not sure it's worth doing -- it seems like at that point you'd be writing specialized ops in C++ anyway.

**Comment by vadimkantorov** [comment]
I thought that given a mask pack/unpack are precisely equivalent to SIMD compress/expand? Aren’t they?

**Comment by ezyang** [comment]
We haven't looked! You're probably right :)

**Comment by vadimkantorov** [comment]
int4 support request in https://github.com/pytorch/pytorch/issues/33859 is also related

**Comment by vadimkantorov** [comment]
Another useful functionality would be scatter/gather-like functionality for compressing index tensors.

In a practical usecase it can help to compress the hybrid sparse+dense tensor indices by a lot: https://discuss.pytorch.org/t/sparse-torch-topk/71832/4

**Comment by vadimkantorov** [comment]
It seems that <8bit quantization starts to appear: https://github.com/pytorch/pytorch/pull/34783/files and seems somewhat related to this discussion

**Comment by ezyang** [comment]
cc @jspark1105

**Comment by vadimkantorov** [comment]
related: https://github.com/pytorch/pytorch/issues/36380

**Comment by vadimkantorov** [comment]
I renamed to enlarge the scope a little bit :) BitTensors could be very helpful for binary neural networks. Even if few operators on them are supported (such as bit packbits/unpackbits, binary operations, popcnt), they are already useful for reducing memory footprint, e.g. for storing masks instead of full inputs when sufficient for backward ops. E.g. in https://github.com/pytorch/pytorch/pull/41034 if a mask is stored, the silu/swish operation would become bijective if additional bit is stored to represent direction (half-space) away from function minimum.

**Comment by vadimkantorov** [comment]
I made a draft:
```python
import math
import torch

def tensor_slice(tensor, dim, s):
    if dim < 0:
        dim += tensor.dim()
    return tensor[(slice(None),) * dim + (s, )]

def packshape(mask, dtype = torch.uint8):
    nbits_element = torch.iinfo(dtype).bits
    nbits = 1 if mask == 0b00000001 else 2 if mask == 0b00000011 else 4 if mask == 0b00001111 else 8 if mask == 0b11111111  else None
    assert nbits is not None and nbits <= nbits_element and nbits_element % nbits == 0
    packed_size = nbits_element // nbits

    return packed_size, nbits

def packbits(tensor, dim = -1, mask = 0b00000001, out = None, dtype = torch.uint8):
    packed_size, nbits = packshape(mask = mask, dtype = dtype)
    shape = list(tensor.shape)
    shape[dim] = int(math.ceil(tensor.size(dim) / packed_size))
    out = out.zero_() if out is not None else torch.zeros(shape, device = tensor.device, dtype = dtype)
    assert list(out.shape) == list(shape)

    for e in range(packed_size):
        sliced_input = tensor_slice(tensor, dim, slice(e, None, packed_size))
        compress = (sliced_input << (nbits * (packed_size - 1 - e)))
        sliced_output = out.narrow(dim, 0, sliced_input.shape[dim])
        sliced_output |= compress

    return out

  def unpackbits(tensor, dim = -1, mask = 0b00000001, out = None, dtype = torch.uint8, shape = None):
    packed_size, nbits = packshape(mask = mask, dtype = tensor.dtype)
    out = out.zero_() if out is not None else torch.zeros(shape, device = tensor.device, dtype = dtype)
    assert list(out.shape) == list(shape)

    for e in range(packed_size):
        sliced_output = tensor_slice(out, dim, slice(e, None, packed_size))
        expand = (tensor >> (nbits * (packed_size - e - 1))) & ((1 << nbits) - 1)
        sliced_input = expand.narrow(dim, 0, sliced_output.shape[dim])
        sliced_output.copy_(sliced_input)

    return out

if __name__ == '__main__':
    shape = (10, 17)
    K = 10
    for nbits in [1, 2, 4, 8]:
        mask = (1 << nbits) - 1
        for dtype in [torch.uint8, torch.int32, torch.int64]:
            for k in range(K):
                x = torch.randint(0, 1 << nbits, shape, dtype = dtype)
                y = packbits(x, mask = mask)
                z = unpackbits(y, mask = mask, dtype = x.dtype, shape = x.shape)
                assert torch.allclose(x, z)
                                             
```

**Comment by vadimkantorov** [comment]
Any advice on how to fuse this properly? Should just torch.jit.script work?

**Comment by vadimkantorov** [comment]
An efficient version for padded tensors so that the relevant dim is multiple of 8 for BoolTensors, multiple of 4 for 2-bit-valued tensors:

```python
def packbits_padded(tensor, dim = -1, mask = 0b1, out = None, dtype = torch.uint8):
    dim = dim if dim >= 0 else dim + tensor.dim()
    nbits_element, nbits = 8, (1 if mask == 0b00000001 else 2 if mask == 0b00000011 else 4 if mask == 0b00001111 else 8 if mask == 0b11111111 else None)
    nibbles = nbits_elmement // nbits
    assert tensor.shape[dim] % nibbles == 0
    
    out = out if out is not None else torch.empty(*tensor.shape[:dim], tensor.shape[dim] // nibbles, *tensor.shape[1 + dim:], dtype = dtype, device = tensor.device)
    shift = torch.arange(nbits_element - nbits, -1, -nbits, dtype = torch.uint8, device = tensor.device)
    shift = shift.view(nibbles, *((1,) * len(tensor.shape[1 + dim:])))
    torch.sum(tensor.view(*tensor.shape[:dim], -1, nibbles, *tensor.shape[1 + dim:]) << shift , dim = -1, out = out)

def unpackbits_padded(tensor, dim = -1, mask = 0b1, out = None):
    dim = dim if dim >= 0 else dim + tensor.dim()
    nbits_element, nbits = 8, (1 if mask == 0b00000001 else 2 if mask == 0b00000011 else 4 if mask == 0b00001111 else 8 if mask == 0b11111111 else None)
    nibbles = nbits_elmement // nbits
    
    out = out if out is not None else torch.empty(*tensor.shape[:dim], tensor.shape[dim] * nibbles, *tensor.shape[1 + dim:], dtype = torch.uint8, device = tensor.device)
    shift = torch.arange(nbits_element - nbits, -1, -nbits, dtype = torch.uint8, device = tensor.device)
    shift = shift.view(nibbles, *((1,) * len(tensor.shape[1 + dim:])))
    torch.bitwise_and((tensor.unsqueeze(1 + dim) >> shift).view_as(out), mask, out = out)
```

**Comment by ezyang** [comment]
If the bitwise ops you need are supported by the fuser, yeah, the JIT fuser should be able to speed some of the things up. I didn't read your code so I don't know for certain if this is the case or not.

**Comment by vadimkantorov** [comment]
Is it possible to ask the fuser to do loop unrolling? (hopefully may exploit op reordering / parallelism)

Practically speaking, I'm using the padded version now which runs much faster.

Some problems encountered with JIT: `shape = list(out.shape)` didn't work. `out.shape[:-1] + (1,)` didn't work in JIT either. iinfo also didn't work, `slice(None)` didn't work. |=, >>= didn't work

**Comment by ezyang** [comment]
^ @suo

**Comment by vadimkantorov** [comment]
about shapes: `(shape[:dim] + (int(math.ceil(shape[dim] / nibbles)), ) + shape[1 + dim:])` does not JIT compile:

```
  File "packbits.py", line 10, in <module>
    def packshape(shape, dim : int = -1, mask : int = 0b00000001, dtype = torch.uint8):
  File "/miniconda/lib/python3.7/site-packages/torch/jit/__init__.py", line 1290, in script
    fn = torch._C._jit_script_compile(qualified_name, ast, _rcb, get_default_args(obj))
RuntimeError:
Arguments for call are not valid.
The following variants are available:

  aten::add.Tensor(Tensor self, Tensor other, *, Scalar alpha=1) -> (Tensor):
  Expected a value of type 'Tensor' for argument 'other' but instead found type 'Tuple[int]'.

  aten::add.Scalar(Tensor self, Scalar other, Scalar alpha=1) -> (Tensor):
  Expected a value of type 'number' for argument 'other' but instead found type 'Tuple[int]'.

  aten::add.out(Tensor self, Tensor other, *, Scalar alpha=1, Tensor(a!) out) -> (Tensor(a!)):
  Expected a value of type 'Tensor' for argument 'other' but instead found type 'Tuple[int]'.

  aten::add(str a, str b) -> (str):
  Expected a value of type 'str' for argument 'a' but instead found type 'Tensor'.

  aten::add.t(t[] a, t[] b) -> (t[]):
  Could not match type Tensor to List[t] in argument 'a': Cannot match List[t] to Tensor.

  aten::add.int(int a, int b) -> (int):
  Expected a value of type 'int' for argument 'b' but instead found type 'Tuple[int]'.

  aten::add.float(float a, float b) -> (float):
  Expected a value of type 'float' for argument 'b' but instead found type 'Tuple[int]'.

  aten::add.int_float(int a, float b) -> (float):
  Expected a value of type 'float' for argument 'b' but instead found type 'Tuple[int]'.

  aten::add.float_int(float a, int b) -> (float):
  Expected a value of type 'int' for argument 'b' but instead found type 'Tuple[int]'.

  aten::add(Scalar a, Scalar b) -> (Scalar):
  Expected a value of type 'number' for argument 'b' but instead found type 'Tuple[int]'.

  add(float a, Tensor b) -> (Tensor):
  Expected a value of type 'Tensor' for argument 'b' but instead found type 'Tuple[int]'.

  add(int a, Tensor b) -> (Tensor):
  Expected a value of type 'Tensor' for argument 'b' but instead found type 'Tuple[int]'.

The original call is:
  File "packbits.py", line 15
    assert bits is not None and nibble is not None and nibble <= bits and bits % nibble == 0
    nibbles = bits // nibble
    return (shape[:dim] + (int(math.ceil(shape[dim] / nibbles)), ) + shape[1 + dim:]), nibbles, nibble
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
```

```
    out = out if out is not None else torch.empty(*tensor.shape[:dim], tensor.shape[dim] // nibbles, *tensor.shape[1 + dim:], dtype = dtype, device = tensor.device)
                                                  ~~~~~~~~~~~~~~~~~~ <--- HERE
    shift = torch.arange(bits - nibble, -1, -nibble, dtype = torch.uint8, device = tensor.device)
    shift = shift.view(nibbles, *((1, ) * (tensor.dim() - dim - 1)))

```

**Comment by vadimkantorov** [comment]
bit packing could also be used for storing masks for backward (dropout, relu, leaky_relu) instead of full input for many operations

**Comment by vadimkantorov** [comment]
@aayn numpy's packbkits is quite restricted and supports packing only BoolTensors. I propose here to expand its functionalities to custom masks and arbitrary data types.

Need for fully-fledged BitTensors also needs some discussion. Same for relation with existing 2-bit quantization and quantization in general, since BitTensors and low-precision quantization could be used for binary neural net works.

**Comment by vadimkantorov** [comment]
More 2-bit packing related to quantization: https://github.com/pytorch/pytorch/pull/43077/files

**Comment by vadimkantorov** [comment]
4-bit tensor dtype: https://github.com/pytorch/pytorch/pull/44574

**Comment by vadimkantorov** [comment]
torch.sum and torch.count_nonzero for bool types should be equivalent. is it the case? Do they allocate any intermediate tensors and upcast? In the ideal case, the should also not upcast too much, if the maximum sum possible fits into uint16 or uint32

POPCNT would be the best though :)

**Comment by vak** [comment]
yet another case, where int8 (and not packed booleans) could have made better use of the very valuable GPU RAM:

```
import numpy as np
import torch
from torch import from_numpy

sz = 2900*1024**2
print(f"sz={sz/1024**2:.1f}MB") # ==> sz=2900.0MB

tensor_vcf = np.ones(sz, dtype=np.bool) # it's just 2.9GB, but you can't sum it up into int32 straight-forward using 15 GPU RAM of Google Colab

device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
cuda1_vcf = from_numpy(tensor_vcf).to(device)

cuda1_vcf.sum(dtype=torch.int32) # ==>  RuntimeError: CUDA out of memory. Tried to allocate 11.33 GiB (GPU 0; 14.73 GiB total capacity; 2.83 GiB already allocated; 10.96 GiB free; 2.83 GiB reserved in total by PyTorch)
```

**Comment by vadimkantorov** [comment]
@vak maybe file a separate issue about this

**Comment by blueardour** [comment]
Mark. Have been searching for such a function for several months. Above code https://github.com/pytorch/pytorch/issues/32867#issuecomment-662619393 works. Expecting more efficient implmentation.

**Comment by ezyang** [comment]
bumping prio based on user activity

**Comment by vadimkantorov** [comment]
I think there are several issues about packing / unpacking, low-bit types in general (and ops for them - maybe different low-bit type representations could exist as well...) and about more support for bit operations (such as popcnt and maybe some others required for binary neural nets)

**Comment by vadimkantorov** [comment]
May be also relevant to https://www.deepspeed.ai/tutorials/onebit-adam/

**Comment by vadimkantorov** [comment]
This is also useful for image region superpixel representations

**Comment by vadimkantorov** [comment]
Probably detectron2 would also make use of this in BitMasks (https://detectron2.readthedocs.io/en/latest/modules/structures.html#detectron2.structures.BitMasks)

**Comment by ezyang** [comment]
I'm working on a feature that would make it easier for people to develop new dtypes out of tree.

**Comment by vadimkantorov** [comment]
Related: https://github.com/pytorch/pytorch/pull/65545

**Comment by vadimkantorov** [comment]
Another useful binary-related function: run-length encoding (RLE), used in segmentation for compressed storage of instances (https://github.com/cocodataset/cocoapi/blob/master/PythonAPI/pycocotools/mask.py)

**Comment by ezyang** [comment]
Quantization has added handling for 1/2/4 bit dtypes, but they didn't handle general striding for these cases.  Here is their size formula:

```
int64_t get_sub_byte_tensor_size(IntArrayRef sizes, size_t dtype_itemsize, at::ScalarType t) {
  // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
  int64_t element_per_byte;
  switch(t) {
    case at::ScalarType::QUInt4x2:
      element_per_byte = 2;
      break;
    case at::ScalarType::QUInt2x4:
      element_per_byte = 4;
      break;
    default:
      element_per_byte = 1;
  }
  // zero dim tensor
  if (sizes.size() == 0) {
    return c10::multiply_integers(sizes) * dtype_itemsize;
  }
  // Consider most inner dim as cols
  int64_t cols = sizes.at(sizes.size()-1);
  int64_t bytes_per_row = cols * dtype_itemsize;
  // align qtensor most inner dim, compute ceil (bytes_per_row / element_per_byte)
  return c10::multiply_integers(IntArrayRef(sizes.data(), sizes.size() - 1)) * at::ceil_div(bytes_per_row, element_per_byte);
}
```

The assumption seems to be that the innermost dim gets packed and rounded and then the outer dims are normal. This means certain restridings are invalid, in particular all strides must maintain the innermost dimension being contiguous.

**Comment by vadimkantorov** [comment]
Would be nice to have the most basic functions: compress booltensor into bittensor and decompress. A pilot case could be saving the dropout mask as bittensor for memory saving

**Comment by vadimkantorov** [comment]
> case at::ScalarType::QUInt4x2:
>       element_per_byte = 2;
>       break;
>     case at::ScalarType::QUInt2x4:
>       element_per_byte = 4;
>       break;
>     default:
>       element_per_byte = 1;

it seems that 1-bit isn't supported here. it would correspond to `element_per_byte = 8`, right?

**Comment by ezyang** [comment]
Oops, you're right!

**Comment by vadimkantorov** [comment]
Related: https://github.com/pytorch/pytorch/issues/74627

**Comment by vadimkantorov** [comment]
Another real-world usecase: packbits was useful for InstantNGP implementation by @kwea123: https://github.com/kwea123/ngp_pl/blob/05c96e314175d1de75a468cddbc264202cef5b81/models/csrc/packbits.cu

**Comment by vadimkantorov** [comment]
this would also be useful as output dtype of torch.bernoulli bool tensor generation, e.g. for dropout masks: https://gist.github.com/vadimkantorov/360ece06de4fd2641fa9ed1085f76d48#file-reludropoutinplace-py-L13

**Comment by ezyang** [comment]
I think the most practical route to bittensors is to add enough bitwise primitives on int (of some size) tensors as necessary, implement operations as Python decompositions, and then rely on torchinductor to compile them into reasonable kernels.

**Comment by vadimkantorov** [comment]
There is also an aspect that for binary ops operating on a float tensor and on bittensor, some sort of cached reads might need to be implemented because of different granularity - a single read of the underlying memory element of bittensor can correspond to many elements of the float tensor (and I guess this unpacking can be done in multiple ways - the simplest for CUDA can probably be doing nothing special and performing this memory read multiple times in the hope that it's going to be cached. but for CPU and making use of masked AVX ops, vectorized reads can be done)

**Comment by vadimkantorov** [comment]
Related: https://github.com/pytorch/vision/issues/6818#issuecomment-1295163095

**Comment by vadimkantorov** [comment]
popcnt may be useful as allocating the result, but also as inplace in rare cases

**Comment by kenseehart** [comment]
I'm quite surprised this is on the back burner. Seems like it should be very high priority. The use cases mentioned seem awfully specific. I'd suggest the use case is any model who's input and/or output is binary data, transfer of that data is a bottleneck, and for which an order of magnitude speed increase is desirable.

Data transfer from CPU to GPU is frequently a bottleneck, so encoding each bit as a 32 bit int or float seems unfortunate.

Also, it is crucial that the unpacking happens on the GPU. Anything else would be a waste.

**Comment by vadimkantorov** [comment]
now each element in BoolTensor gets encoded as 8 bit, but otherwise I agree that implementing more BitTensor functionality will enable less wasteful copies/storage and more experimentation

**Comment by vadimkantorov** [comment]
https://github.com/pytorch/pytorch/pull/95860 was merged in, so maybe some functions like packbits/unpackbits can be thought of. Also, maybe some plan could be thought of for use of bits type in place of BoolTensor for some places in core (like maybe some masks stored as saved_tensors for autograd)

**Comment by ezyang** [comment]
Yeah, in my ideal world we start building out some kernel support, leveraging Inductor for it. If anyone is interested in pitching in I can help scope out exactly what this would involve.

**Comment by vadimkantorov** [comment]
in addition to packbits, some other rudimentary de/quantize can be implemented given that llama.cpp made 4-bit quantization more popular and practical

**Comment by Felix-Petersen** [comment]
The `difflogic` library provides a fast CUDA kernel implementation for packbits that operates on PyTorch tensors: https://github.com/Felix-Petersen/difflogic

**Comment by vadimkantorov** [comment]
Very interesting! I think core would be happy to have the basic  kernels for packbits/unpackbits at least

**Comment by Felix-Petersen** [comment]
My current (unreleased) version of packing and unpacking is:

### `CUDA`
```cpp
#define CHECK_CUDA(x) TORCH_CHECK(x.type().is_cuda(), #x " must be a CUDA tensor")
#define CHECK_CONTIGUOUS(x) TORCH_CHECK(x.is_contiguous(), #x " must be contiguous")
#define CHECK_INPUT(x)                                                                                                 \
    CHECK_CUDA(x);                                                                                                     \
    CHECK_CONTIGUOUS(x)

// adapted from https://stackoverflow.com/questions/14038589/what-is-the-canonical-way-to-check-for-errors-using-the-cuda-runtime-api
#define gpuErrchk(ans)                                                                                                 \
    { gpuAssert((ans), __FILE__, __LINE__); }
inline void gpuAssert(const cudaError_t code, const char *const file, const int line, const bool abort = true) {
    if (code != cudaSuccess) {
        fprintf(stderr, "GPUassert: %s %s %d\n", cudaGetErrorString(code), file, line);
        if (abort)
            exit(code);
    }
}

template <typename T> T ceil_div(const T x, const T y) { return x / y + !!(x % y); }

/**********************************************************************************************************************/

template <typename scalar_t>
__global__ void tensor_packbits_cuda_kernel(
    torch::PackedTensorAccessor64<bool, 2, torch::RestrictPtrTraits> t,
    torch::PackedTensorAccessor64<scalar_t, 2, torch::RestrictPtrTraits> b
) {

    for (
        auto row = blockIdx.y * blockDim.y + threadIdx.y;
        row < b.size(0);
        row += blockDim.y * gridDim.y
    ) {
        for (
            auto col = blockIdx.x * blockDim.x + threadIdx.x;
            col < b.size(1);
            col += blockDim.x * gridDim.x
        ) {

            typedef typename std::make_unsigned<scalar_t>::type unsigned_scalar_t;
            union {
                unsigned_scalar_t unsigned_scalar;
                scalar_t signed_scalar;
            } val;
            constexpr int bit_count = std::numeric_limits<unsigned_scalar_t>::digits;
            val.signed_scalar = b[row][col];
            for (unsigned int i = 0; i < bit_count; ++i) {
                const unsigned_scalar_t bit_mask = static_cast<unsigned_scalar_t>(t[row][bit_count * col + i]) << i;
                val.unsigned_scalar = val.unsigned_scalar | bit_mask;
            }
            b[row][col] = val.signed_scalar;
        }
    }
}

std::tuple<torch::Tensor, int> tensor_packbits_cuda(
    torch::Tensor t,
    const int bit_count
) {
    CHECK_INPUT(t);

    const auto batch_in_size = t.size(1);
    const auto batch_out_size = ceil_div(batch_in_size, static_cast<int64_t>(bit_count));
    const auto out_size = t.size(0);
    const auto pad_len = (bit_count - batch_in_size % bit_count) % bit_count;

    dim3 threads_per_block(32, 32);

    const dim3 blocks_per_grid(
        min(static_cast<int64_t>(65535), ceil_div(batch_out_size, static_cast<int64_t>(threads_per_block.x))),
        min(static_cast<int64_t>(65535), ceil_div(out_size, static_cast<int64_t>(threads_per_block.y)))
    );

    auto dispatch_type = [bit_count]() {
        switch (bit_count) {
        case 8:
            return torch::kInt8;
        case 16:
            return torch::kInt16;
        case 32:
            return torch::kInt32;
        case 64:
            return torch::kInt64;
        default:
            throw std::invalid_argument("`bit_count` has to be in { 8, 16, 32, 64 }");
        }
    }();
    auto b = torch::zeros({out_size, batch_out_size}, torch::dtype(dispatch_type).device(t.device()));

    AT_DISPATCH_INTEGRAL_TYPES(b.type(), "tensor_packbits_cuda_kernel", ([&] {
                                   tensor_packbits_cuda_kernel<scalar_t><<<blocks_per_grid, threads_per_block>>>(t.packed_accessor64<bool, 2, torch::RestrictPtrTraits>(),
                                                                                                                 b.packed_accessor64<scalar_t, 2, torch::RestrictPtrTraits>());
                               }));
    gpuErrchk(cudaPeekAtLastError());
    gpuErrchk(cudaDeviceSynchronize());

    return {b, pad_len};
}


/**********************************************************************************************************************/


template <typename scalar_t>
__global__ void tensor_unpackbits_cuda_kernel(
    torch::PackedTensorAccessor64<scalar_t, 2, torch::RestrictPtrTraits> t,
    torch::PackedTensorAccessor64<bool, 2, torch::RestrictPtrTraits> b,
    const int bit_count
) {

    for (
        auto row = blockIdx.y * blockDim.y + threadIdx.y;
        row < b.size(0);
        row += blockDim.y * gridDim.y
    ) {
        for (
            auto col = blockIdx.x * blockDim.x + threadIdx.x;
            col < b.size(1);
            col += blockDim.x * gridDim.x
        ) {

            const auto bit = (t[row][col / bit_count] >> (col % bit_count)) & 1;
            b[row][col] = static_cast<bool>(bit);

        }
    }
}

torch::Tensor tensor_unpackbits_cuda(
    torch::Tensor t,
    const int bit_count,
    const int pad_len
) {
    CHECK_INPUT(t);

    const auto batch_in_size = t.size(1);
    const auto batch_out_size = batch_in_size * bit_count - pad_len;
    const auto out_size = t.size(0);

    dim3 threads_per_block(32, 32);

    const dim3 blocks_per_grid(
        min(static_cast<int64_t>(65535), ceil_div(batch_out_size, static_cast<int64_t>(threads_per_block.x))),
        min(static_cast<int64_t>(65535), ceil_div(out_size, static_cast<int64_t>(threads_per_block.y)))
    );

    auto b = torch::zeros({out_size, batch_out_size}, torch::dtype(torch::kBool).device(t.device()));

    AT_DISPATCH_INTEGRAL_TYPES(t.type(), "tensor_unpackbits_cuda_kernel", ([&] {
                                   tensor_unpackbits_cuda_kernel<scalar_t><<<blocks_per_grid, threads_per_block>>>(t.packed_accessor64<scalar_t, 2, torch::RestrictPtrTraits>(),
                                                                                                                   b.packed_accessor64<bool, 2, torch::RestrictPtrTraits>(),
                                                                                                                   bit_count
                                                                                                                   );
                               }));
    gpuErrchk(cudaPeekAtLastError());
    gpuErrchk(cudaDeviceSynchronize());

    return b;
}

```

### `CPP`
```cpp
std::tuple<torch::Tensor, int> tensor_packbits_cuda(
    torch::Tensor t,
    const int bit_count
);
torch::Tensor tensor_unpackbits_cuda(
    torch::Tensor t,
    const int bit_count,
    const int pad_len
);

PYBIND11_MODULE(TORCH_EXTENSION_NAME, m) {
    m.def(
        "tensor_packbits_cuda",
        [](torch::Tensor t, const int bit_count) {
            return tensor_packbits_cuda(t, bit_count);
        },
        "tensor_packbits_cuda (CUDA)");
    m.def(
        "tensor_unpackbits_cuda",
        [](torch::Tensor t, const int bit_count, const int pad_len) {
            return tensor_unpackbits_cuda(t, bit_count, pad_len);
        },
        "tensor_unpackbits_cuda (CUDA)");
}
```

### `Python`
```python
def pack_along_dim_1(t, bit_count=64, device='cuda'):
    assert len(t.shape) == 2, t.shape
    assert device == 'cuda', device

    t = t.contiguous().to(device)

    return difflogic_cuda.tensor_packbits_cuda(t, bit_count)

def unpack_along_dim_1(t, pad_len, device='cuda'):
    assert len(t.shape) == 2, t.shape
    assert device == 'cuda', device

    t = t.contiguous().to(device)

    return difflogic_cuda.tensor_unpackbits_cuda(t, torch.iinfo(t.dtype).bits, pad_len)

if __name__ == '__main__':
    # Test pack_along_dim_1 / unpack_along_dim_1

    # t = torch.rand(123, 3910).round().bool()
    t0 = torch.rand(128, 2097152).round().bool().cuda()
    print(t0.shape, t0.dtype)
    t, pad = pack_along_dim_1(t0, 64)
    print(t.shape, t.dtype, pad)
    t2 = unpack_along_dim_1(t[::20], pad)
    print(t2.shape, t2.dtype)
    t = unpack_along_dim_1(t, pad)
    print(t.shape, t.dtype)

    print((t0 == t).float().mean())
    assert (t0 == t).all()
```

**Comment by vadimkantorov** [comment]
Matlab's function for packing binary images into uint32: https://www.mathworks.com/help/images/ref/bwpack.html

**Comment by vadimkantorov** [comment]
Maybe `torch.bits8` or `torch.bits16` can be used as target for such pack/unpack functions. It would also be nice to support such bits tensor creation with passed Python array of bools (`torch.tensor([True]*8, dtype = torch.bits8)` gives currently `RuntimeError: invalid type`) and maybe `.tolist()` to such a python list of bools, only endianness needs to be decided or e.g. only big endian could be supported for such packing)

**Comment by vadimkantorov** [comment]
Do you think it's better to have a separate tensor dtype for the semantics "compressed bool tensor" or just ab/use torch.bits8 for this? I guess the important question is how to treat indexing `bool_tensor[4]` versus `bit_tensor[4]` (maybe some special function can be implemented on bittensor/bits8 for this?) and how to treat cases if original BoolTensor shapes do not divide 8 (one option might be to simply not support this or to always pad and let the user save somewhere else the original BoolTensor's shape). Then it would be nice to use it for saving the mask in old eager's dropout impl and allow fused impls like `torch.gt(float_tensor, 4, dtype = torch.bits)`

**Comment by ezyang** [comment]
So we just agreed internally that we are going to implement a uint1 type. So you can use that. Note that sub byte dtypes need a lot of extra infra which we plan to implement using Python tensor subclasses

**Comment by vadimkantorov** [comment]
I think, for being useful packbits/unpackbits + bit ops + some fused kernels that currently produce only BoolTensor are okay for already being useful. Going forward, it might be nice to have all things that currently support torch.bool to also support torch.uint1. Btw, if we have some generic tiled dtypes, dispatch system for sub-byte dtypes can also become simpler as they are naturally tiled (e.g. think usecase for torch.gt(float_tensor, 4, dtype = torch.bitmap) where a tile of 8 floats could ideally be read at once and be processed into a single byte (tile of 8 bits) on the output). These are probably already dispatched in the similar way for the vectorized kernels on CPU? (as tiled vec types exist)

Also, I must bikeshed-confess that uint1 is super unintuitive name for the high-level bitmap/bitmask/bittensor concept. Maybe worth naming it `bitmap`? or `bit`/`bits`?

For subclass, I think it would be good to have auto-upcast to torch.bool, so that existing bool kernels can consume bitmaps right away, hiding a reallocation for the eager mode (and ideally optimized away by inductor for the compiled case? )

**Comment by vadimkantorov** [comment]
torch.bit is in thanks to @jerryzh168 https://github.com/pytorch/pytorch/pull/117208 :)

**Comment by Felix-Petersen** [comment]
I might be missing the actual changes, but as I understand it https://github.com/pytorch/pytorch/pull/117208 only high-level declares the existence of uint1 etc. Am I misunderstanding this?

**Comment by Felix-Petersen** [comment]
BTW, here is a PyTorch (not native CUDA) implementation of arbitrary k-bit bitpacking (into int64s) in the context of quantization: 

https://github.com/Felix-Petersen/distquant/blob/dcd5a94825d9022543b3e95afa91e5590cd72f2d/distquant/distquant.py#L262

https://github.com/Felix-Petersen/distquant/blob/dcd5a94825d9022543b3e95afa91e5590cd72f2d/distquant/distquant.py#L305

**Comment by ezyang** [comment]
Yeah none of the actual implementation exists yet. Also, the implementation is going to all be in Python, so to get low overhead performance you are going to have to use torch.compile eventually

**Comment by Felix-Petersen** [comment]
Would we really expect that torch.compile will be able to perform efficient bitpacking? Single bit accesses seem something that ideally should be optimized on a very low level (eventually potentially instruction level). At the end of the day, a lot depends on the specifics of memory accesses, but, for any low bit cases, we need to access at least 8 bits for reading a single bit, and memory address-wise, the data type should probably not exist outside of 1+ dimensional tensors. I believe it requires significant consideration to support it properly. (And hardware considerations should be taken into account if it becomes part of a stable release.) 

Just to clarify, the plan is still that we would still be storing 8 bits in 1 byte (contrasting the current 1 bit in 1 byte of bool).

**Comment by ezyang** [comment]
> Just to clarify, the plan is still that we would still be storing 8 bits in 1 byte (contrasting the current 1 bit in 1 byte of bool)?

Yes.

> Would we really expect that torch.compile will be able to perform efficient bitpacking? Single bit accesses seem something that ideally should be optimized on a very low level (eventually potentially instruction level). At the end of the day, a lot depends on the specifics of memory accesses, but, for any low bit cases, we need to access at least 8 bits for reading a single bit, and memory address-wise, the data type should probably not exist outside of 1+ dimensional tensors.

So, at least in the short term the way I would expect to implement these operators in terms of regular bitwise operations on the bit tensor reinterpreted as a uint8_t (or larger) tensor. The planned implementation for these sub-byte tensors doesn't really allow for non-aligned accesses anyway (see https://github.com/albanD/subclass_zoo/blob/main/uint4_tensor.py for PoC).

**Comment by vadimkantorov** [comment]
Also, it could be good to support many ops relevant for torch.bit tensors (bit ops / pack ops etc) on all tensors as well using reinterpret, but encourage/advertise their use only on torch.bit examples (e.g. in https://github.com/pytorch/pytorch/issues/105465 it might make sense to support bit flips on float tensors as well as it could be used for flipping sign or on setting LSB to some external value).

At the end of the day, IMHO, what matters is existence of these ops with nice, clear, specific naming and description. So maybe forcing a subclass or reinterpret could even be avoided, although the op examples could promote use of torch.bit for bitset/bitmap/compressedbooltensor usecases

**Comment by vadimkantorov** [comment]
> Also, the implementation is going to all be in Python

I guess it would be good to have benchmarks in test comparing at least initially these Python-compiled bit ops with some manual CUDA loops to be confident in them, as unrolling multiple loop iterations related to bit to have a single memory store per byte (or even several bytes) might be a non-trivial optimization. Maybe for this it would be good to have these wide-dtypes in Python as well (e.g. like float32x8, then it would  be sth like: read 8-tuple of floats from a float32tensor.view(torch.float32x8) and then store a single byte as produced output as a single store. or maybe even wider dtypes as outputs (and inputs) would be a good abstraction to help torch.compile produce effective code for these kinds of patterns (read, compute/pack/reduce, write)

**Comment by vadimkantorov** [comment]
Regarding sub-byte access, I would propose to treat this by introducing some variants of getbit/setbit/togglebit ops available for any dtype: https://github.com/pytorch/pytorch/issues/105465, at least this would unblock and make access more elegant and error-free.

**Comment by LemonPi** [comment]
To chime in, my use case for this is to efficiently represent occupancy voxel grids where each voxel is either occupied or not (don't need 8 bits info). I'm running into memory issues when batching these voxel grids, and so better memory efficiency would be great.

However, these voxel grids are also frequently read from/written to, so performance is also a concern when considering packing/unpacking.

**Comment by vadimkantorov** [comment]
Somewhat related (although used ternary) - BitNet: https://arxiv.org/abs/2402.17764

**Comment by vadimkantorov** [comment]
Maybe given the revival of interest in extremely low-bit quantization methods, maybe would be nice to also include some eager methods in core (mainly for correct tested semantics, not for actualy speed - speed / Linear op support may come later). I personally am more in favor of adding them first as eager methods working with dense/bits tensors without adding tensor subclasses. Now there is a lot of experimentation with quantization methods, it's developing very fast and unclear if any particular method wins. A lof of this experimentation is working with very simple Linear module swaps, and for this having some eager method under the hood and passing around a few tensors manually representing qparams is completely fine.

**Comment by vadimkantorov** [comment]
Related on pack/unpack tuples of floats: https://github.com/pytorch/pytorch/issues/125129. It would be nice to have various performant pack/unpack bit utils in core PyTorch (both in eager for semantics experiments + for fusion with triton codegen)

**Comment by msaroufim** [comment]
@vadimkantorov as a baseline we were thinking of writing bit packs in pure pytorch and codegening efficient kernels for them here https://github.com/pytorch/ao/pull/291 - it's a good baseline and we'll look to see how those kernels could be made faster

**Comment by vadimkantorov** [comment]
I think it's important to have a manual kernel as a baseline for perf comparison at least for some cases to be sure that the generated code is fast enough (or at least to review it / have some tests so that it doesn't evolve to something inefficient with Inductor evolvement and development)

**Comment by vadimkantorov** [comment]
> as a baseline

maybe also worth supporting the original numpy's API as well: np.packbits/np.unpackbits - which could redirect to the more general, versatile interface
